### PR TITLE
Add support for serializing shapes on the body with jsonvalue members.

### PR DESCRIPTION
**Pull request by alextwoods**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

1. To make sure we include your contribution in the release notes, please make sure to add description entry for your changes in the "unreleased changes" section of the `CHANGELOG.md` file (at corresponding gem). For the description entry, please make sure it lives in one line and starts with `Feature` or `Issue` in the correct format.

2. For generated code changes, please checkout below instructions first:
  https://github.com/aws/aws-sdk-ruby/blob/version-3/CONTRIBUTING.md

Thank you for your contribution!

### CODE REVIEW

File: gems/aws-sdk-core/lib/aws-sdk-core/param_validator.rb [hunk 1]
```diff
@@ -152,7 +152,7 @@ def shape(ref, value, errors, context)
       when MapShape then map(ref, value, errors, context)
       when DocumentShape then document(ref, value, errors, context)
       when StringShape
-        unless value.is_a?(String)
+        unless value.is_a?(String) || ref['jsonvalue']
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/param_validator.rb]: I don't think this one is needed. The jsonvalue converter should run before this.

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb [hunk 2]
```diff
@@ -11,15 +11,31 @@ class Handler < Seahorse::Client::Handler
...
-              param_value = context.params[m]
-              unless param_value.respond_to?(:to_json)
-                raise ArgumentError, "The value of params[#{m}] is not JSON serializable."
+            convert_jsonvalue(m, ref, context.params)
+          end
+          @handler.call(context)
+        end
+
+        def convert_jsonvalue(m, ref, params)
+          return if params.nil? || params[m].nil?
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb]: We want to allow for serializing nil if it's provided explicitly. I think we will want to check params.key?(m) and if present, serialize it. Also can params ever be nil? It should always be at least a hash.

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb [hunk 3]
```diff
@@ -11,15 +11,31 @@ class Handler < Seahorse::Client::Handler
...
+
+          if ref['jsonvalue']
+            param_value = params[m]
+            unless param_value.respond_to?(:to_json)
+              raise ArgumentError, "The value of params[#{m}] is not JSON serializable."
+            end
+            params[m] = param_value.to_json
+          elsif ref.shape.is_a?(Seahorse::Model::Shapes::StructureShape)
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb]: I think we need to handle other shape types, like Map and Union?

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb (hunk 2 above)
Comment (alextwoods) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb]: The params can be nil here only because its getting called recursively.  But I think I agree that we should use .key? and serialize a nil that is explicitly set.

File: gems/aws-sdk-core/lib/aws-sdk-core/param_validator.rb (hunk 1 above)
Comment (alextwoods) [reply gems/aws-sdk-core/lib/aws-sdk-core/param_validator.rb]: Yup - removed.

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb (hunk 3 above)
Comment (alextwoods) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb]: I'm not 100% positive which shapes we need to support - but I think you're right, I think anything with members that can reference a string shape.  I'll update.

Comment (mullermp) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb]: Better yet, is there a solution that doesn't involve shape checks here?

Comment (alextwoods) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb]: The only other option (which is what I started with) was to check `respond_to?(:members)` and `respond_to?(:member)` and `respond_to?(:value)`.  But I think thats worse than just being shape based.

Comment (mullermp) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb]: Yeah, agreed, then shape is fine, assuming we cover them all.

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb (hunk 2 above)
Comment (mullermp) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb]: Yes. And we need a test for not serializing "null" and doing so when deliberate.

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb [hunk 4]
```diff
@@ -11,15 +11,43 @@ class Handler < Seahorse::Client::Handler
 
         def call(context)
           context.operation.input.shape.members.each do |m, ref|
-            if ref['jsonvalue']
-              param_value = context.params[m]
-              unless param_value.respond_to?(:to_json)
-                raise ArgumentError, "The value of params[#{m}] is not JSON serializable."
+            convert_jsonvalue(m, ref, context.params, 'params')
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb:14]: I think we can just call this method and let the "structure shape" case handle the top level input?

Comment (alextwoods) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb:14]: I think so - but will require a refactoring to remove the `m` parameter.  But should be possible.

Comment (alextwoods) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/jsonvalue_converter.rb:14]: Edit: This doesn't quite work - because we need to set values back on the params it gets tricky.  The code ends up being a bit cleaner here if we keep the top level iteration of input members.

**Review by mullermp — APPROVED** [review]
LGTM apart from formatting / line lengths in specs.
